Lib/asyncio/unix_events.py

```diff
@@ -878,6 +878,73 @@ def __exit__(self, a, b, c):
         raise NotImplementedError()
 
 
+class PidfdChildWatcher(AbstractChildWatcher):
+    """Child watcher implementation using Linux's pid file descriptors.
+
+    This child watcher polls process file descriptors (pidfds) to await child
+    process termination. In some respects, PidfdChildWatcher is a "Goldilocks"
+    child watcher implementation. It doesn't require signals or threads, doesn't
+    interfere with any processes launched outside the event loop, and should
+    scale linearly with the number of subprocesses launched by the event
+    loop. The main disadvantage is that pidfds are specific to Linux, and only
+    work on recent (5.3+) kernels.
+    """
+
+    def __init__(self):
+        self._loop = None
+        self._callbacks = {}
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, a, b, c):
+        pass
+
+    def is_active(self):
+        return self._loop is not None and self._loop.is_running()
+
+    def close(self):
+        self.attach_loop(None)
```

Comment:
IMO, it seems like `close()` should clear the callbacks rather than `attach_loop()`. 
```suggestion
        self._callbacks.clear()
        self.attach_loop(None)
```

This is the current behavior in all three of the other child watchers that use a callbacks dict:

`SafeChildWatcher`:
https://github.com/python/cpython/blob/5c0c325453a175350e3c18ebb10cc10c37f9595c/Lib/asyncio/unix_events.py#L959-L960

`FastChildWatcher`:
https://github.com/python/cpython/blob/5c0c325453a175350e3c18ebb10cc10c37f9595c/Lib/asyncio/unix_events.py#L1038-L1039

`MultiLoopChildWatcher`:
https://github.com/python/cpython/blob/5c0c325453a175350e3c18ebb10cc10c37f9595c/Lib/asyncio/unix_events.py#L1153-L1154

Reply:
As far as I'm aware, the examples you list don't keep state in the event loop. Closing a `PidfdChildWatcher` should remove it completely from the loop.